pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -48,54 +57,59 @@ type Reconciler struct {
 }
 
 func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
-	sts := &appsv1.StatefulSet{}
-	err := r.client.Get(ctx, req.NamespacedName, sts)
-	if err != nil {
-		// we'll ignore not-found errors, since there is nothing to do.
-		return ctrl.Result{}, client.IgnoreNotFound(err)
-	}
-
-	log := ctrl.LoggerFrom(ctx).WithValues("statefulset", klog.KObj(sts))
+	log := ctrl.LoggerFrom(ctx).WithValues("statefulset", klog.KRef(req.Namespace, req.Name))
 	ctx = ctrl.LoggerInto(ctx, log)
 	log.V(2).Info("Reconciling StatefulSet")
 
-	err = r.fetchAndFinalizePods(ctx, sts)
-	if err != nil {
-		return ctrl.Result{}, err
-	}
-
-	return ctrl.Result{}, nil
+	err := r.fetchAndFinalizePods(ctx, req)
+	return ctrl.Result{}, err
 }
 
-func (r *Reconciler) fetchAndFinalizePods(ctx context.Context, sts *appsv1.StatefulSet) error {
+func (r *Reconciler) fetchAndFinalizePods(ctx context.Context, req reconcile.Request) error {
 	podList := &corev1.PodList{}
-	if err := r.client.List(ctx, podList, client.InNamespace(sts.Namespace), client.MatchingLabels{
-		podcontroller.GroupNameLabel: GetWorkloadName(sts.Name),
+	if err := r.client.List(ctx, podList, client.InNamespace(req.Namespace), client.MatchingLabels{
+		podcontroller.GroupNameLabel: GetWorkloadName(req.Name),
 	}); err != nil {
 		return err
 	}
+
+	sts := &appsv1.StatefulSet{}
+	err := r.client.Get(ctx, req.NamespacedName, sts)
```

Comment:
I think this does not work well if Kueue is restarted after the reconcile event was started and before the Reconciler happened. 

The similar problem is solved in [k8s Job controller](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/job/job_controller.go) which also needs to clean finalizers of orphan pods and is resilient to restarts. We use the "ADDED" event to Pods to delete the orhaned pods. PTAL.

The solution in k8s might be hard to implement, but I think in Kueue we could relatively easliy add an additional recnciler for Pods owned by statefulSet. When the StatefulSet is gone then finalize the Pods.

Reply:
> I think this does not work well if Kueue is restarted after the reconcile event was started and before the Reconciler happened.

You're right, I missed handling create events. Fixed.

Reply:
I also tested locally. It is working perfectly fine.